### DISCUSSION TITLE
Access to upstreamPublisher/upstreamSubscriber in subscriptions

### DISCUSSION THREAD

**Discussion by richardbizik** [graphql-java/graphql-java#2466] [Q&A] [answered]
Hi, would it be possible to add getter for upstreamPublisher into CompletionStageMappingPublisher 
https://github.com/graphql-java/graphql-java/blob/d783ec7ee30369ca2fdecdfd0b74e8bf6d96baa8/src/main/java/graphql/execution/reactive/CompletionStageMappingPublisher.java#L25
as well as use non anonymous class for Subscriber with getter for  https://github.com/graphql-java/graphql-java/blob/d783ec7ee30369ca2fdecdfd0b74e8bf6d96baa8/src/main/java/graphql/execution/reactive/CompletionStageMappingPublisher.java#L41
and another getter for DelegatingSubscription
https://github.com/graphql-java/graphql-java/blob/d783ec7ee30369ca2fdecdfd0b74e8bf6d96baa8/src/main/java/graphql/execution/reactive/DelegatingSubscription.java#L13
so we could have access to our underlying implementation. I have some info stored in these classes and executing 
``` java
GraphQL graphQL = GraphQL
        .newGraphQL(graphQLSchema)
        .instrumentation(instrumentation)
        .build();
ExecutionResult executionResult = graphQL.execute(executionInput);
executionResult.getData();
```
returns instance of CompletionStageMappingPublisher It would be useful to have access to my data stored in upstream. Or is there something I am missing and the access is already possible?

**Comment by bbakerman** [comment, marked as answer]
Originally I said - I take that back on more reflection

> Yeah raise a PR - I see no reason why we should not support this.

> The general contract is that it returns a Publisher of events - but if you going to cast it to a specific instance - so be it.

> The risk is that this could change - but its unlikely.

**Reply by bbakerman** [reply]
Although having a second closer look at this, I think this will not be that great.

The class is marked as @Internal to reduce out API surface area (I missed that when I looked earlier)

Can I ask more on what you need access to   - after all you gave out the Publisher in the first place.

As a work around you could use GraphQL context to pass it back outwards

I would love to get more details on your use case

**Reply by richardbizik** [reply]
I have a thread running under DataFetcher<Publisher> producing changes of documents, this thread also holds references to subscribers to which these changes should be published. My issue is that I can populate the list of subscribers with new ones because I can pass the list down to publisher returned by DataFetcher:get method and create new publisher with the reference to the list and add a new instance of subscriber in Subscriber:onSubscribe method but no way to remove subscribers from said list in DataFetcher instance
with the proposed changes I would be able to do something like: 
Class handling graphql subscription
```
GraphQL graphQL = GraphQL
        .newGraphQL(graphQLSchema)
        .instrumentation(instrumentation)
        .build();
ExecutionResult executionResult = graphQL.execute(executionInput);
Publisher<ExecutionResult> stream = executionResult.getData();
CompletionStageMappingPublisher publisher = (CompletionStageMappingPublisher) stream;
QueryEventPublisher upstreamPublisher = (QueryEventPublisher) publisher.getUpstreamPublisher(); <<< Get my instance of the publisher
QuerySubscriber subscriber = new QuerySubscriber();
stream.subscribe(subscriber);
subscriber.onSubscribe(new QuerySubscription(upstreamPublisher::unsubscribe, subscriber));
```
and in QuerySubscription.class I would be able to remove subscriber from my list by calling cancel on the subscription
```
public QuerySubscription(Consumer<Subscriber> unsubscribe, QuerySubscriber subscriber) {
    this.unsubscribe = unsubscribe;
    this.subscriber = subscriber;
}
@Override
public void cancel() {
    unsubscribe.accept(subscriber);
}
```
and in QueryEventPublisher I would be able to create a method that would be able to remove subscribers from my list
```
public void unsubscribe(Subscriber s) {
    subscribers.removeIf(subscriber -> ((CompletionStageMappingPublisher.CompletionStageSubscriber)subscriber).getDownstreamSubscriber().equals(s));
}
```

**Reply by bbakerman** [reply]
Just remember there is 2 steps in this subscription Publisher - on each event it pushes out new publisher events.

But also it applies the graphql query to the event object

```
  onEvent {
      selection1
      selection2
 }
 ```

so the underlying publisher has the raw events but then they are then turned into selection set mapped objects via the CompletionStageMappingPublisher mapper

You dont want to lose this by delving too deep - you might hit a Balrog

**Reply by richardbizik** [reply]
Yes I am pushing events to the top level `CompletionStageMappingPublisher.CompletionStageSubscriber` so I can keep all that functionality and I am using reference to be able to delete the instance from my list by checking which one holds the instance of `Subscriber` I want to delete.

**Reply by bbakerman** [reply]
See https://github.com/graphql-java/graphql-java/pull/2474 for a greater API garuantee

**Reply by richardbizik** [reply]
That looks great. Thanks
